src/dune_rules/fetch_rules.ml

```diff
@@ -169,19 +204,21 @@ let gen_rules_for_checksum_or_url url checksum =
   let rules =
     Rules.collect_unit
     @@ fun () ->
+    let* url = resolve_url url in
     (* CR-rgrinberg: it's possible to share the downloading step between the
        directory and file actions. Though it's unlikely to be of any use in real
        world situations. *)
     let rule =
-      let info = Rule.Info.of_loc_opt (Some (fst url)) in
+      let info = Rule.Info.of_loc_opt (Some loc_url) in
       fun { Action_builder.With_targets.build; targets } ->
         Rules.Produce.rule (Rule.make ~info ~targets build)
     in
     let make_target = make_target checksum_or_url in
     let action ~target ~kind =
-      action ~url ~checksum ~target ~kind
+      action ~url:(loc_url, url) ~checksum ~target ~kind
       |> Action.Full.make
-      |> Action_builder.With_targets.return
+      |> Action_builder.return
+      |> Action_builder.with_no_targets
```

Comment:
I'm trying to understand, why is this without targets?

Reply:
They the targets are added in the functions below. It's done separately because there's either a file or a directory target.